File: ocaml/plugins/infisical.ml
(* The `infisical` provider kind, as a voxgig/plugin definition.

   The Infisical API over HTTPS, with universal auth as the source of a
   token.

   A port of typescript/plugins/infisical.ts, which is canonical. *)

open Secret
open Provider
open Httpjson

(* Infisical.

   `api.token` reads the secret keyed `API_TOKEN` (Infisical's own
   convention is environment-style keys) at a secret path in one environment
   of a project. Auth is a token, or a universal-auth (machine identity)
   login. Its expiry field is `expiresIn`, camelCase, unlike everyone
   else's `expires_in`. *)
let infisical_provider (addr : string) (token : string) (clientid : string)
    (clientsecret : string) (project : string) (environment : string) (secretpath : string) :
    provider =
  let livetoken = ref None in
  let renewat = ref never in

  let login useaddr =
    if "" <> token then token
    else begin
      if "" = clientid || "" = clientsecret then
        fail "sekreto: infisical: no token and no client credentials";

      let body =
        Json.obj [ ("clientId", Json.str clientid); ("clientSecret", Json.str clientsecret) ]
      in

      let res =
        fetchjson
          ~headers:[ ("content-type", "application/json") ]
          ~body:(Json.stringify body) "POST"
          (useaddr ^ "/api/v1/auth/universal-auth/login")
      in

      let got = digtext res.jbody [ "accessToken" ] in
      if 200 <> res.status || not (isset got) then
        fail ("sekreto: infisical login failed: " ^ string_of_int res.status);

      renewat := renewtime (Json.odig res.jbody [ "expiresIn" ]);
      Option.get got
    end
  in

  {
    lookup =
      (fun name ->
        let useaddr = trimslash (firstof [ addr; "https://app.infisical.com" ]) in
        checkaddr useaddr;

        if "" = project || "" = environment then fail "sekreto: infisical: no project/environment";

        if None = !livetoken || nowms () >= !renewat then livetoken := Some (login useaddr);

        let url =
          useaddr ^ "/api/v3/secrets/raw/" ^ envkey name ^ "?workspaceId="
          ^ Sigv4.uriescape project ^ "&environment=" ^ Sigv4.uriescape environment
          ^ "&secretPath=" ^ Sigv4.uriescape (firstof [ secretpath; "/" ])
        in

        let res =
          fetchjson
            ~headers:[ ("authorization", "Bearer " ^ Option.value ~default:"" !livetoken) ]
            "GET" url
        in

        if 404 = res.status then None
        else if 200 <> res.status then fail ("sekreto: infisical error: " ^ string_of_int res.status)
        else digtext res.jbody [ "secret"; "secretValue" ]);
    describe = (fun () -> "infisical:" ^ project ^ "/" ^ environment);
  }

let plugin () : Defs.definition =
  providerplugin "infisical" (fun spec ->
      infisical_provider spec.addr spec.token spec.clientid spec.clientsecret spec.project
        spec.environment spec.secretpath)

File: cpp/plugins/Infisical.cpp
#include "Infisical.hpp"

#include "Httpjson.hpp"
#include "Providers.hpp"
#include "Sekreto.hpp"

namespace sekreto {

namespace {

// -------------------------------------------------------------- infisical

/// Infisical.
///
/// `api.token` reads the secret keyed `API_TOKEN` (Infisical's own
/// convention is environment-style keys) at a secret path in one
/// environment of a project. Auth is a token, or a universal-auth (machine
/// identity) login with clientid/clientsecret.
class InfisicalProvider : public Provider {
 public:
  InfisicalProvider(const std::string& addr, const std::string& token,
                    const std::string& clientid, const std::string& clientsecret,
                    const std::string& project, const std::string& environment,
                    const std::string& path)
      : addr_(addr),
        token_(token),
        clientid_(clientid),
        clientsecret_(clientsecret),
        project_(project),
        environment_(environment),
        path_(path) {}

  std::optional<std::string> lookup(const std::string& name) override {
    std::string useaddr = trimslash(first(addr_, "https://app.infisical.com"));
    checkaddr(useaddr);

    if (project_.empty() || environment_.empty()) {
      throw SekretoError("sekreto: infisical: no project/environment");
    }

    if (!livetoken_.has_value() || nowms() >= renewat_) {
      livetoken_ = login(useaddr);
    }

    // envkey here takes NO prefix.
    std::string url = useaddr + "/api/v3/secrets/raw/" + envkey(name) +
                      "?workspaceId=" + uriescape(project_) +
                      "&environment=" + uriescape(environment_) +
                      "&secretPath=" + uriescape(first(path_, "/"));

    Ordered headers;
    headers.set("authorization", "Bearer " + livetoken_.value_or(""));

    Answer res = fetchjson("GET", url, headers);

    if (404 == res.status) return std::nullopt;

    if (200 != res.status) {
      throw SekretoError("sekreto: infisical error: " + std::to_string(res.status));
    }

    return textof(res.body.get("secret").get("secretValue"));
  }

  std::string describe() const override {
    return "infisical:" + project_ + "/" + environment_;
  }

 private:
  std::string login(const std::string& useaddr) {
    if (!token_.empty()) return token_;

    if (clientid_.empty() || clientsecret_.empty()) {
      throw SekretoError("sekreto: infisical: no token and no client credentials");
    }

    Json payload = Json::obj({{"clientId", Json::str(clientid_)},
                              {"clientSecret", Json::str(clientsecret_)}});

    Ordered headers;
    headers.set("content-type", "application/json");

    Answer res = fetchjson("POST", useaddr + "/api/v1/auth/universal-auth/login", headers,
                           Json::stringify(payload));

    std::optional<std::string> got = textof(res.body.get("accessToken"));

    if (200 != res.status || !got.has_value() || got.value().empty()) {
      throw SekretoError("sekreto: infisical login failed: " + std::to_string(res.status));
    }

    // camelCase, unlike everyone else's expires_in.
    renewat_ = renewtime(res.body.get("expiresIn"));

    return got.value();
  }

  std::string addr_;
  std::string token_;
  std::string clientid_;
  std::string clientsecret_;
  std::string project_;
  std::string environment_;
  std::string path_;

  std::optional<std::string> livetoken_;
  double renewat_ = NEVER;
};

}  // namespace

Definition infisical() {
  return providerplugin("infisical", [](const ProviderSpec& spec) -> std::shared_ptr<Provider> {
    return std::make_shared<InfisicalProvider>(spec.addr, spec.token, spec.clientid,
                                               spec.clientsecret, spec.project,
                                               spec.environment, spec.path);
  });
}

}  // namespace sekreto
